lib/active_merchant/billing/gateways/iveri.rb

```diff
@@ -0,0 +1,231 @@
+        super
+      end
+
+      def purchase(money, payment_method, options={})
+        post = build_vxml_request('Debit', options) do |xml|
+          add_params_in_order(xml, :purchase, money, payment_method, options)
+        end
+
+        commit(post)
+      end
+
+      def authorize(money, payment_method, options={})
+        post = build_vxml_request('Authorisation', options) do |xml|
+          add_params_in_order(xml, :authorize, money, payment_method, options)
+        end
+
+        commit(post)
+      end
+
+      def capture(money, authorization, options={})
+        post = build_vxml_request('Debit', options) do |xml|
+          add_params_in_order(xml, :capture, money, authorization, options)
+        end
+
+        commit(post)
+      end
+
+      def refund(money, authorization, options={})
+        post = build_vxml_request('Credit', options) do |xml|
+          add_params_in_order(xml, :refund, money, authorization, options)
+        end
+
+        commit(post)
+      end
+
+      def void(authorization, options={})
+        post = build_vxml_request('Void', options) do |xml|
+          add_params_in_order(xml, :void, nil, authorization, options)
+        end
+
+        commit(post)
+      end
+
+      def verify(credit_card, options={})
+        authorize(0, credit_card, options)
+      end
+
+      def verify_credentials
+        auth = authorize(0, '', options)
+        return true if auth.error_code == '14'
+        false
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((CertificateID=\\\")[^\\]*), '\1[FILTERED]').
+          gsub(%r((&lt;PAN&gt;)[^&]*), '\1[FILTERED]').
+          gsub(%r((&lt;CardSecurityCode&gt;)[^&]*), '\1[FILTERED]')
+      end
+
+      private
+
+      def build_xml_envelope(vxml)
+        builder = Nokogiri::XML::Builder.new(:encoding => 'UTF-8') do |xml|
+          xml[:soap].Envelope 'xmlns:xsi' => 'http://www.w3.org/2001/XMLSchema-instance', 'xmlns:xsd' => 'http://www.w3.org/2001/XMLSchema', 'xmlns:soap' => 'http://schemas.xmlsoap.org/soap/envelope/' do
+            xml[:soap].Body do
+              xml.Execute 'xmlns' => 'http://iveri.com/' do
+                xml.validateRequest 'true'
+                xml.protocol 'V_XML'
+                xml.protocolVersion '2.0'
+                xml.request vxml
+              end
+            end
+          end
+        end
+
+        builder.to_xml
+      end
+
+      def build_vxml_request(action, options)
+        builder = Nokogiri::XML::Builder.new do |xml|
+          xml.V_XML('Version' => '2.0', 'CertificateID' => @options[:cert_id], 'Direction' => 'Request') do
+            xml.Transaction('ApplicationID' => @options[:app_id], 'Command' => action, 'Mode' => mode) do
+              yield(xml)
+            end
+          end
+        end
+
+        builder.doc.root.to_xml
+      end
+
+      def add_params_in_order(post, action, money, pm, options)
```

Comment:
After some sifting I was able to break things down into a more idiomatic structure with fewer conditionals. The potential downside is it may be less easy to extend in the future if we need to fit in additional request fields or modes. Updating the PR...